DependencyInjection/Compiler/HashGeneratorPass.php

```diff
@@ -34,15 +34,30 @@ public function process(ContainerBuilder $container)
 
         $definition = $container->getDefinition('fos_http_cache.user_context.hash_generator');
 
+        $prioritisedTags = array();
+        $taggedProviders = $container->findTaggedServiceIds(self::TAG_NAME);
+
+        if (!count($taggedProviders)) {
+            throw new InvalidConfigurationException('No user context providers found. Either tag providers or disable fos_http_cache.user_context');
+        }
+
+        foreach ($taggedProviders as $id => $tags) {
+            foreach ($tags as $tag) {
+                $priority = isset($tag['priority']) ? (int)$tag['priority'] : 0;
+                $prioritisedTags[$priority][] = $id;
+            }
+        }
+
+        krsort($prioritisedTags, SORT_NUMERIC);
+        $prioritisedProviders = call_user_func_array('array_merge', $prioritisedTags);
+
         $providers = array();
-        foreach ($container->findTaggedServiceIds(self::TAG_NAME) as $id => $parameters) {
+        foreach ($prioritisedProviders as $id) {
             $providers[] = new Reference($id);
         }
 
-        if (!count($providers)) {
-            throw new InvalidConfigurationException('No user context providers found. Either tag providers or disable fos_http_cache.user_context');
+        if (!empty($providers)) {
```

Comment:
how could providers be empty now? you kept the sanity check with `count` above, so i don't think we need this check.